org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/launch/BaseLaunchRequestJob.java

```diff
@@ -46,24 +47,37 @@ public Void runInToolingApi(CancellationTokenSource tokenSource, IProgressMonito
     }
 
     protected final void executeLaunch(CancellationTokenSource tokenSource, final IProgressMonitor monitor) throws Exception {
-        // todo (etst) close streams when done
-
-        BuildExecutionParticipants.activateParticipantPlugins();
+        BuildExecutionParticipants buildParticipants = BuildExecutionParticipants.create(CorePlugin.extensionManager().loadBuildExecutionParticipants());
+        
         monitor.beginTask(getJobTaskName(), IProgressMonitor.UNKNOWN);
 
         ProcessDescription processDescription = createProcessDescription();
         RunConfiguration runConfig = getRunConfig();
+        
         InternalGradleBuild gradleBuild = CorePlugin.internalGradleWorkspace().getGradleBuild(runConfig.getProjectConfiguration().getBuildConfiguration());
+
+        buildParticipants.preAction(runConfig.getProjectConfiguration().getProjectDir(), runConfig.getTasks(), monitor);
+
         GradleProgressAttributes attributes = GradleProgressAttributes.builder(tokenSource, monitor)
-                .forDedicatedProcess(processDescription)
-                .withFullProgress()
-                .build();
+            .forDedicatedProcess(processDescription)
+            .withFullProgress()
+            .build();
         T launcher = createLaunch(gradleBuild, runConfig, attributes, processDescription);
 
         writeExtraConfigInfo(attributes);
 
         Event event = new DefaultExecuteLaunchRequestEvent(processDescription, launcher);
         CorePlugin.listenerRegistry().dispatch(event);
+       
+        launcher.addProgressListener(new ProgressListener() {
+            @Override
+            public void statusChanged(ProgressEvent event) {
+                if (event instanceof TaskFinishEvent) {
+                    buildParticipants.postAction(runConfig.getProjectConfiguration().getProjectDir(),
```

Comment:
buildParticipants should be able to see when build is finished, no matter if the TaskFinished or not right?

Reply:
you mean every progressEvent will trigger postBuild action?